### ISSUE TITLE
Jpmml Predicts different than SparkML for XGBoost

### ISSUE THREAD

**Issue by fatihtekin** [jpmml/jpmml-sparkml-xgboost#1]
When I use XGBoost jpmml it gives different result than spark xgboost model for the same record.   

  "org.apache.spark" %% "spark-core" % "2.2.0",
  "org.apache.spark" %% "spark-sql" % "2.2.0",
  "org.apache.spark" %% "spark-mllib" % "2.2.0" exclude("org.jpmml", "pmml-model"),
  "ml.dmlc" % "xgboost4j" % "0.7",
  "ml.dmlc" % "xgboost4j-spark" % "0.7",
  "org.jpmml" % "jpmml-sparkml" % "1.3.2",
  "org.jpmml" % "jpmml-sparkml-xgboost" % "1.0-SNAPSHOT",
  "org.jpmml" % "jpmml-xgboost" % "1.2-SNAPSHOT",
  "org.jpmml" % "pmml-evaluator" % "1.3.8"

```
import java.io.ByteArrayOutputStream
import java.util
import javax.xml.bind.JAXBException
import ml.dmlc.xgboost4j.scala.spark.{XGBoostClassificationModel, XGBoostEstimator}
import org.apache.spark.ml.estimator.veon.RareMerger
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator
import org.apache.spark.ml.feature.{OneHotEncoder, SQLTransformer, StringIndexer, VectorAssembler}
import org.apache.spark.ml.tuning.{CrossValidator, ParamGridBuilder}
import org.apache.spark.ml.{Pipeline, PipelineStage}
import org.apache.spark.sql.SparkSession
import org.dmg.pmml.{FieldName, PMML}
import org.jpmml.evaluator.{CategoricalValue, ContinuousValue, FieldValue, ModelEvaluatorFactory}
import org.jpmml.model.MetroJAXBUtil
import org.jpmml.sparkml.ConverterUtil
import scala.collection.mutable

object JPMMLPipeline {
  def main(args: Array[String]): Unit = {
    val spark = SparkSession.builder().master("local[*]").appName("Trusted Payments Mleap").config("spark.driver.memory","12g").getOrCreate()
    var train = spark.read.format("csv").option("header", "true").option("inferSchema", "true").load("Train.csv").repartition(8)
    var test = spark.read
      .format("csv") .option("header", "true")  header.option("inferSchema", "true").load("Test.csv").repartition(8)
    val categorical_columns = Array("List of Categories")
    var indexers = mutable.ListBuffer[PipelineStage]()
    categorical_columns foreach(categoryColumn =>  {
      val indexer = new StringIndexer()
        .setInputCol(categoryColumn )
        .setOutputCol(categoryColumn + "_INDEX")
        .setHandleInvalid("keep")
      indexers += indexer
    })
    var encoderColumns = mutable.ListBuffer[String]()
    var encoders = mutable.ListBuffer[PipelineStage]()
    categorical_columns foreach(categoryColumn =>  {
      val encoder = new OneHotEncoder()
        .setInputCol(categoryColumn+"_INDEX")
        .setOutputCol(categoryColumn + "_VEC")
        .setDropLast(false)
      encoderColumns += categoryColumn + "_VEC"
      encoders += encoder
    })
    val features = train.columns.filter(!_.contains("label")).filter(!categorical_columns.contains(_))
    val assembler = new VectorAssembler().setInputCols(features ++ encoderColumns).setOutputCol("features")

    val pipeline = new Pipeline().setStages((indexers ++ encoders ++ Array(assembler)).toArray)
    val trainModel = pipeline.fit(train)
    val paramMap = List(
      "eta" -> 0.1,
      "max_depth" -> 7,
      "objective" -> "binary:logistic",
      "num_round" ->10,
      "eval_metric" -> "auc",
      "nworkers" -> 8).toMap
    val xgboostEstimator = new XGBoostEstimator(paramMap)
    val pipelineXGBoost = new Pipeline().setStages(Array(trainModel, xgboostEstimator))
    val cvModel = pipelineXGBoost.fit(train)
    println(cvModel.stages(1).asInstanceOf[XGBoostClassificationModel].booster.getModelDump(null,true,"text")(0))
    val pmml = ConverterUtil.toPMML(train.schema, cvModel)
    val evalTest = test.limit(1)
    evalTest.show(1, true)
    val modelResult = cvModel.transform(evalTest)
    modelResult.select("probabilities", "prediction").show(1,false)
    val modelEvaluatorFactory = ModelEvaluatorFactory.newInstance
    val jpmmlEvaluator = modelEvaluatorFactory.newModelEvaluator(pmml)
    val arguments = new util.LinkedHashMap[FieldName, FieldValue]()

    val sparkToJpmmlMap :Map[org.apache.spark.sql.types.DataType, org.dmg.pmml.DataType]= Map(
      org.apache.spark.sql.types.IntegerType -> org.dmg.pmml.DataType.INTEGER,
      org.apache.spark.sql.types.DoubleType -> org.dmg.pmml.DataType.DOUBLE,
      org.apache.spark.sql.types.StringType -> org.dmg.pmml.DataType.STRING)
    val evalRecord = evalTest.take(1)(0)
    evalRecord.schema.zipWithIndex.foreach { case (field, i) =>
      if(categorical_columns.contains(field.name)){
        arguments.put(new FieldName(field.name), CategoricalValue.create(sparkToJpmmlMap(field.dataType), evalRecord.get(i)))
      }else{
        arguments.put(new FieldName(field.name), ContinuousValue.create(sparkToJpmmlMap(field.dataType), evalRecord.get(i)))
      }
    }
    val pmmlResults = jpmmlEvaluator.evaluate(arguments)
    println(pmmlResults)
  }
}
```

**Comment by vruusmann** [comment]
This is a known issue - categorical features, where the number of categories is greater than 2, are not supported.

My troubleshooting efforts indicate that the problem is located on the XGBoost4J side, not JPMML-SparkML/JPMML-Evaluator side. In other words, XGBoost4J creates invalid DMatrix for some categorical features.

Please experiment with more toy datasets, and confirm the following:
1. Predictions are correct when the feature set contains unlimited number of continuous features.
2. Predictions are correct when the feature set contains unlimited number of continuous and binary categorical features.
3. Predictions are all wrong when the feature set contains a single multi-class categorical feature.

The JPMML-SparkML-XGBoost project has integration tests, which all pass cleanly:
https://github.com/jpmml/jpmml-sparkml-xgboost/tree/master/src/test/resources

If you have time/interest to troubleshoot the internals of XGBoost4J, then go ahead. I have other more pressing responsibilities at the moment.

**Comment by vruusmann** [comment]
Never do `(Continuous|Categorical)Value.create(sparkToJpmmlMap(field.dataType), evalRecord.get(i))` in your code - you're bypassing PMML data preparation logic (eg. the missing value replacement logic is not executed), which may lead to bad predictions. Always do `org.jpmml.evaluator.InputField#prepare(Object)`.

Anyway, the above is not a likely cause for the above-stated problem. There is something broken inside XGBoost4J.

**Comment by fatihtekin** [comment]
Do you have the issue link to XGBoost4j?

**Comment by vruusmann** [comment]
No, I haven't formalized my observation as a GitHub issue yet.

I have experimented with a toy dataset (a label column, and a multi-class categorical feature column), and XGBoost4J predictions and PMML predictions do not match. I'm 99.78% sure that PMML predictions are the correct ones in this case (multi-class categorical features are fully supported when executing XGBoost directly, or via R and Scikit-Learn wrappers; things fall apart only when using the XGBoost4J wrapper).

**Comment by fatihtekin** [comment]
Thank you so much. Could you please give an update in this ticket if you create a ticket in XGboost4j project?

**Comment by fatihtekin** [comment]
Hi 
I have tried to use no categorical column and dropped all of them from the dataframe. So if I only have continuous columns still getting different values

I set to categorical_columns = Array[String]() . and dropped all the categorical columns from the dataset.


```
    val evalTest = test.limit(1)
    val modelResult = cvModel.transform(evalTest)
    modelResult.select("probabilities", "prediction").show(1,false)

    val modelEvaluatorFactory = ModelEvaluatorFactory.newInstance
    val jpmmlEvaluator = modelEvaluatorFactory.newModelEvaluator(pmml)
    val arguments = new util.LinkedHashMap[FieldName, FieldValue]()

    val evalRecord = evalTest.take(1)(0)
    import scala.collection.JavaConverters._

    val fieldMap: Map[String, InputField] = jpmmlEvaluator.getInputFields().asScala.map(inputField => inputField.getName.getValue -> inputField).toMap

    evalRecord.schema.zipWithIndex.foreach { case (field, i) =>
      if(fieldMap.contains(field.name)){
        arguments.put(fieldMap(field.name).getName, fieldMap(field.name).prepare(evalRecord.getAs(field.name)))
      }
    }
    evalTest.show(1, true)
    println(arguments)
    val pmmlResults = jpmmlEvaluator.evaluate(arguments)
    println(pmmlResults)

```

**Comment by vruusmann** [comment]
The source of problems is the model generation-side of code (Apache Spark ML pipeline for data preprocessing + XGBoost4J for model training + JPMML-SparkML-XGBoost for model export). For example, if you make a mistake encoding continuous/categorical information properly in Apache Spark ML pipeline, then all the subsequent steps will see bad data, and the exported PMML document will be bad as well.

You're currently showing me your model evaluations-side code, which looks fine. However, if you working with a badly exported PMML document, then no model evaluation-side code change can really fix the problem.  

So, using the Iris dataset (which only contains continuous features), please show me how you get "bad predictions".

**Comment by vruusmann** [comment]
In other words, the PMML document that you're trying to execute using the JPMML-Evaluator library contains bad prediction logic. 

Your need to fix your PMML document, not "troubleshoot" the JPMML-Evaluator library.

**Comment by fatihtekin** [comment]
I don't understand why would that cause the model predict different results. You can see in my first comment how I am preparing the data. I am only using standard SparkML features, no custom estimator/translator.  With the last test, I only drop all the categorical columns and only left continuous columns in the dataset and still getting conflicting results of predictions.

**Comment by vruusmann** [comment]
The JPMML-XGBoost library has always done correct job with categorical features.

All mismatching predictions can be traced back to bad application code at data developer's end. Specifically, one has to be very careful when passing one-hot-encoded columns from Apache Spark to XGBoost4J, because the latter interprets `1` (category active), `0` (category inactive) and `Double.NaN` (there is nothing known about a category) very differently.

Here's my integration test about categorical features: https://github.com/jpmml/jpmml-sparkml-xgboost/blob/master/src/test/resources/XGBoostAudit.scala
